(feat) KHP3-8729: Add service availability configuration with weekly scheduling

gabriel090 opened the pull request:
## Requirements

- [ ] This PR has a title that briefly describes the work done, including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work is based on designs, which are linked or shown either in the Jira ticket or the description below.
- [ ] My work includes tests or is validated by existing tests.

## Summary
<!-- Please describe what problems your PR addresses. -->
Add a new "Configure Service Availability" feature that allows administrators
to set maximum appointment limits at both the service-level and per-day level through
weekly schedules.
Features Added:
* Global maximum appointment limit for all days
* Weekly availability toggle with day-specific configuration
* Per-day maximum limits (overrides global limit)
* Per-day time ranges (start/end times)
* Days not selected are blocked for appointments

### 1. Set up for Maximum Limit for All Days
**Purpose:** Set a default maximum appointment limit that applies to all days of the week for a service.
**How to Use:**
1. Open "Configure Service Availability" workspace
2. Select a service from the dropdown
3. Enter a number in the "Max Appointments Limit" field
4. If weekly availability is **NOT** enabled, this limit applies to all days
**Behavior:**
- The limit is stored in the `maxAppointmentsLimit` field of the appointment service
- Applies uniformly across all days unless weekly availability overrides it
- Leave empty for unlimited appointments
- Accepts only positive integers


https://github.com/user-attachments/assets/42980491-f9db-4deb-8cef-fce3a5844795



### 2. Booking When Maximum Limit is Reached

**Current Behavior:**
- The system currently **does not** validate against the maximum limit when booking appointments
- The `maxAppointmentsLimit` field is configured and saved, but not enforced during booking

**Expected Behavior (To Be Implemented):**
When a user attempts to book an appointment:
1. The system should check the current appointment count for that service/day
2. Compare against the configured `maxAppointmentsLimit.`
3. If the limit is reached:
   - Display alert: "Maximum appointment limit reached for [Service Name] on [Date]."
   - Prevent booking
   - Suggest alternative dates/times

https://github.com/user-attachments/assets/d71b4007-6a72-4a0c-9595-0af8d8b34b80



### 3. Weekly Setup for a Service

**Purpose:** Configure different availability schedules and limits for each day of the week.

**How to Enable:**
1. Select a service
2. Toggle "Enable weekly availability?" to **Yes**
3. Day selection checkboxes appear for all days of the week

**Selecting Days:**

**Step 1: Check desired days**
- Check any combination of days (Monday through Sunday)
- Multiple days can be selected
**Day-Specific Maximum Limits:**
- Each day can have its own `maxAppointmentsLimit.`
- Overrides the global service limit for that specific day
- Leave empty for unlimited appointments on that day


https://github.com/user-attachments/assets/3d962600-ce55-4eda-a2bd-4b0333527aa3



### 4. How It Works When the Max Limit for a Specific Day is Reached
**Current Implementation:**
- The limit is saved in the database
- Backend stores the configuration in the `weeklyAvailability` array


https://github.com/user-attachments/assets/53b7124e-b257-498d-b6ec-23e665fd1afe




## Screenshots
<!-- Required if you are making UI changes. -->

## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
<!-- https://issues.openmrs.org/browse/O3- -->

## Other
<!-- Anything not covered above -->

gabriel090 committed:
feat(appointments): add service availability configuration with weekly scheduling

makombe commented:
@gabriel090 Please add a ticket in the openmrs Jira for tracking purposes and future reference. Thanks

makombe commented:
@gabriel090 Please add a ticket in the openmrs Jira for tracking purposes and future reference. Thanks

chibongho commented on packages/esm-appointments-app/src/form/appointments-form.workspace.tsx:
```diff
@@ -62,7 +68,10 @@ interface AppointmentsFormProps {
 
 const isValidTime = (timeStr: string) => timeStr.match(new RegExp(time12HourFormatRegexPattern));
-
+const SERVICE_APPOINTMENT_CAPS: Record<string, number> = {
```
Looks like this is no used. Remove

chibongho reviewed (commented):
Not a thorough review; just some initial comments.

gabriel090 commented on packages/esm-appointments-app/src/form/service-availability-form.test.tsx:
```diff
@@ -0,0 +1,362 @@
+
+describe('AppointmentServiceForm', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
```
fixed

gabriel090 commented on packages/esm-appointments-app/src/form/service-availability-form.test.tsx:
```diff
@@ -0,0 +1,362 @@
+  it('renders the service availability form with all fields', async () => {
+    renderWithSwr(<AppointmentServiceForm {...defaultProps} />);
+
+    await waitFor(() => {
```
Fixed. Thanks

gabriel090 commented:
> I second this. Looks like the backend actually supports the feature you're working on. That said, we have not implemented this on the frontend because we haven't needed it. Has that changed? If so, a JIRA to document the requirements would be good.

Yes, the requirements have changed. This feature is now needed to manage appointment capacity and prevent overbooking at our facilities. 
Key requirements:
1. Global appointment limits: Set a maximum number of appointments allowed per day across all days of the week
2. Weekly availability configuration: Enable/disable specific days of the week for appointments
3. Day-specific limits: Override the global limit for individual days (e.g., fewer appointments on Fridays)
4. Day-specific time ranges: Configure different operating hours for each day

gabriel090 committed:
-migrate to Workspace v2 and address PR feedback

- Migrated service availability form to Workspace v2 architecture
- Removed inline styles (opacity) in favor of CSS classes
- Optimized tests: removed clearAllMocks, replaced waitFor with findBy
- Removed comments and unused code from SCSS and TypeScript files

gitcliff commented on packages/esm-appointments-app/src/hooks/useAppointmentsCalendar.ts:
```diff
@@ -66,3 +66,124 @@ function evaluateAppointmentCalendarDates(forDate: string, period: string) {
+function buildUpdatePayload(uuid: string, payload: any, existingService: any): any {
+  const mergedPayload: any = {
+    uuid,
+    name: payload.name,
```
```suggestion
    name: payload.name ?? existingService?.name ?? '',
```

gitcliff commented on packages/esm-appointments-app/src/hooks/useAppointmentsCalendar.ts:
```diff
@@ -66,3 +66,124 @@ function evaluateAppointmentCalendarDates(forDate: string, period: string) {
+  }
+}
+
+function buildUpdatePayload(uuid: string, payload: any, existingService: any): any {
```
thanks @gabriel090 for the work on this , but the function would benefit from stronger typing because right now both payload and `existingService` are typed as any, which removes typeScript’s safety guarantees and makes the function harder to reason about.

Defining interfaces (e.g., `ServicePayload`, `ExistingService`) would help:
prevent silent bugs from incorrect or missing fields ,give better IntelliSense and documentation for future contributors, make refactoring safer when backend models evolve, clarify which fields are optional vs required

Adding these interfaces will make the function more predictable and easier to maintain in the long term.

gabriel090 committed:
Enhance buildUpdatePayload with nullish coalescing

gabriel090 committed:
Merge branch 'main' into KHP3-8729

gabriel090 committed:
Merge branch 'main' into KHP3-8729
